astropy/io/fits/column.py

```diff
@@ -1389,19 +1389,21 @@ def __getattr__(self, name):
 
         Implements for example self.units, self.formats, etc.
         """
-
         cname = name[:-1]
         if cname in KEYWORD_ATTRIBUTES and name[-1] == 's':
             attr = []
-            for col in self:
+            for col in self.columns:
                 val = getattr(col, cname)
-                if val is not None:
-                    attr.append(val)
-                else:
-                    attr.append('')
+                attr.append(val if val is not None else '')
             return attr
         raise AttributeError(name)
 
+    def _update_attrs(self, name=None):
+        """Get or update local lists of column attributes"""
+        attrs = [name] if name is not None else KEYWORD_ATTRIBUTES
+        for name in attrs:
+            setattr(self, name + 's', self.__getattr__(name + 's'))
```

Comment:
>  (if it is not too difficult to find all uses of cols.names[i]).

It seems not so easy to do obviously. Part of the problem is that ColDefs is a list of columns, so to find a column with its name, the `names` attribute has to be accessed from several places (`self._coldefs.names.index(column.name)` in `_convert_other`, or `_get_index(self.names, key)` in `__getitem__`: each time a column is accessed from its name, the list of names is built). So caching these attributes seems the most straightforward solution.